src/inference/hmc.jl

```diff
@@ -676,3 +620,65 @@ end
 
 
 AHMCAdaptor(alg::Hamiltonian) = nothing
+
+##########################
+# HMC State Constructors #
+##########################
+
+function HMCState(model::Model, spl::Sampler{<:StaticHamiltonian}; kwargs...)
+    # Reuse the VarInfo.
+    vi = spl.state.vi
+
+    # Get the metric type.
+    metricT = getmetricT(spl.alg)
+
+    ∂logπ∂θ = gen_∂logπ∂θ(vi, spl, model)
+    logπ = gen_logπ(vi, spl, model)
+    h = AHMC.Hamiltonian(
+        metricT(length(vi[spl])),
+        logπ, ∂logπ∂θ)
+    traj = gen_traj(spl.alg, spl.alg.ϵ)
+
+    return HMCState(vi, 0, 0, traj, h, AHMC.Adaptation.NoAdaptation())
+end
+
+function HMCState(model::Model,
+        spl::Sampler{<:AdaptiveHamiltonian};
+        adaptor=AHMCAdaptor(spl.alg),
+        kwargs...
+)
+    # Reuse the VarInfo.
+    vi = spl.state.vi
+
+    # Link everything if needed.
+    link!(vi, spl)
+
+    # Get the initial log pdf and gradient functions.
+    ∂logπ∂θ = gen_∂logπ∂θ(vi, spl, model)
+    logπ = gen_logπ(vi, spl, model)
+
+    # Get the metric type.
+    metricT = getmetricT(spl.alg)
+
+    # Create a Hamiltonian.
+    θ_init = Vector{Float64}(spl.state.vi[spl])
+    metric = metricT(length(θ_init))
+    h = AHMC.Hamiltonian(metric, logπ, ∂logπ∂θ)
+
+    # Find a step size.
+    init_ϵ = spl.alg.init_ϵ
+
+    # Find good eps if not provided one
+    if init_ϵ == 0.0
```

Comment:
Why these things is done in HMCState instead of the common `sample_init`?

Reply:
To avoid type instability. The stuff in this section requires a `VarInfo` and a `Sampler`, but the `Sampler` I am amble to create before this code completes doesn't represent the typing of the sampler at the end of the code. The trajectories, metrics, Hamiltonian, etc. are all typed in here and then bundled up into a fully parameterized `HMCState` struct. 

I want to find a better way to do this but for right now it works. It really should go away, but at some point the PR has to stop so we can get it into the codebase and all of these design decision are not being made by me entirely.

Reply:
Got it thanks!